### PR TITLE
Nuggets manager to ensure proper (and simpler) nuggets rendering

### PR DESCRIPTION
**Pull request by n0izn0iz**
https://bluenove.atlassian.net/browse/DEVAS-814

### CODE REVIEW

File: assembl/static2/js/app/components/common/nuggetsManager.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  constructor() {
+    this.nuggetsList = [];
+  }
+  add(nuggets: Nuggets) {
+    this.nuggetsList.push(nuggets);
+    this.sort();
+    this.update();
+  }
+  update = () => {
+    let previous = null;
```
Comment (ayyazdaniaryan) [inline assembl/static2/js/app/components/common/nuggetsManager.js]: No no. Please don't use mutations like this. You're using a functional approach, keep to it.
```
this.nuggestList.forEach( (nugget, index, nuggets) => {
  if (index == 0) {
    nugget.updateTop(null);
  }
  nugget.updateTop(nuggets[index-1]);
});

File: assembl/static2/js/app/components/common/nuggetsManager.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    this.nuggetsList.forEach((nuggets) => {
+      nuggets.updateTop(previous);
+      previous = nuggets;
+    });
+  };
+  remove(nuggets: Nuggets) {
+    const index = this.nuggetsList.indexOf(nuggets);
+    if (invalidIndex(index)) throw new Error('Tried to remove nuggets that are not managed by this NuggetsManager');
```
Comment (ayyazdaniaryan) [inline assembl/static2/js/app/components/common/nuggetsManager.js]: Let the error include the index that caused the error.

File: assembl/static2/js/app/components/common/nuggetsManager.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
+// @flow
+
+import Nuggets from '../debate/thread/nuggets';
+
+const invalidIndex = (index) => {
+  return index === -1;
```
Comment (ayyazdaniaryan) [inline assembl/static2/js/app/components/common/nuggetsManager.js]: actually, an invalid index is `index < 0`. But, I understand this case will never be hit in this PR.

File: assembl/static2/js/app/components/debate/thread/post.jsx [hunk 4]
```diff
@@ -194,7 +205,19 @@ class Post extends React.PureComponent {
...
+          postId={id}
+          nuggetsManager={nuggetsManager}
+          completeLevel={(fullLevel
+            ? [
+              rowIndex,
+              ...fullLevel.split('-').map((string) => {
+                return Number(string);
+              })
+            ]
+            : [rowIndex]).join('-')}
```
Comment (ayyazdaniaryan) [inline assembl/static2/js/app/components/debate/thread/post.jsx]: This entire if block is next to unreadable for a developer not very familiar with this part of the code. For the sake of clarity, can it be written outside of the Component creation as a variable and cleaned up?

File: assembl/static2/js/app/components/common/nuggetsManager.js (hunk 3 above)
Comment (n0izn0iz) [reply assembl/static2/js/app/components/common/nuggetsManager.js]: from MDN
>Array.prototype.indexOf()
>....
>Return value
>The first index of the element in the array; -1 if not found.

Comment (ayyazdaniaryan) [reply assembl/static2/js/app/components/common/nuggetsManager.js]: I understand. The method name implies it checks all cases of invalid indices, not just the returned value of `indexOf`. Perhaps change the method name?

Comment (n0izn0iz) [reply assembl/static2/js/app/components/common/nuggetsManager.js]: yup

**Pull request merged**